### PR TITLE
run Doxygen and doc coverage test on static Travis config

### PR DESCRIPTION
**Pull request by 3nids**

### CODE REVIEW

File: cmake/CreateQgsVersion.cmake [hunk 1]
```diff
@@ -0,0 +1,47 @@
+MACRO(CREATE_QGSVERSION)
```
Comment (m-kuhn) [inline cmake/CreateQgsVersion.cmake]: Can you add a header with description of what this does?

File: doc/CMakeLists.txt [hunk 2]
```diff
@@ -46,65 +46,65 @@ IF(WITH_APIDOC)
...
-      ${CMAKE_SOURCE_DIR}/src/gui/raster
-      ${CMAKE_SOURCE_DIR}/src/gui/symbology-ng
-      ${CMAKE_SOURCE_DIR}/src/analysis
-      ${CMAKE_SOURCE_DIR}/src/analysis/interpolation
-      ${CMAKE_SOURCE_DIR}/src/analysis/network
-      ${CMAKE_SOURCE_DIR}/src/analysis/openstreetmap
-      ${CMAKE_SOURCE_DIR}/src/analysis/raster
-      ${CMAKE_SOURCE_DIR}/src/analysis/vector
-      ${CMAKE_SOURCE_DIR}/src/plugins
+      ${CMAKE_CURRENT_SOURCE_DIR}/../doc
```
Comment (m-kuhn) [inline doc/CMakeLists.txt]: Why did you change this to relative paths? I recently changed all the include paths to absolute (relative to `CMAKE_SOURCE_DIR`) and this commit here seems to go in the opposite direction.

Comment (3nids) [inline doc/CMakeLists.txt]: The issue is that this file is included from `.ci/travis/code_layout/CMakeLists.txt` hence the cmake top source directory is not the same.
In general, I would prefer to use the relative path for this reason, it allows more flexibilty.
What do you think?

Comment (m-kuhn) [inline doc/CMakeLists.txt]: It's kind of ok here.

But you soon go 3 levels up and then it requires a lot of brainpower to determine, where the target is (`../../core/`: was it the core directory in src/, python/, tests/? etc.).
Another drawback is, that when adding a new folder (e.g. the subfolder `src/core/expressions`) you can't just copy-paste this between different CMakeLists but you have to manually check for each and every single CMakeLists file the relative path.

Personally, I find these relative paths in CMakeLists a PITA to maintain and would prefer to switch it all to "relative to source root". And move this CMakeLists file here also to the root.

So I wonder if there is a possibility to move the logic from this CMakeLists file here to the main one or move all code related to this one to this subdirectory.

Comment (3nids) [inline doc/CMakeLists.txt]: That means having a config which does not build any QGIS part at all.
That means a huge part of the main CMakeLists would be wrap in something like `IF (WITH_CORE)`
Do you think it's better? 
I would say yes, since we have a single starting point for any build.

Comment (m-kuhn) [inline doc/CMakeLists.txt]: Yes, that's probably what it means. I agree that a single entry point sounds better.

**Comment by 3nids** [comment]
I think this is ready to be merged.

* I added a `WITH_CORE` option which basically allows not to build QGIS. If you use `WITH_ASTYLE`, `WITH_API_DOC` and `ENABLE_TESTS` this will allow the static tests to be run on source code.
* astyle has been moved from `/src/astyle` to `/lib/astyle` (side remark: wouldn't it be better to have all external libs there, with maybe git submodules?)
* `.ci/travis/code_layout/CMakeLists.txt` has been removed since all its functionality were brought into top CMakeLists.txt
* for the moment, sip and doc coverage tests are still run in linux build until checking that new classes have their sip counterparts (ongoing [discussion](https://lists.osgeo.org/pipermail/qgis-developer/2017-June/048740.html))

@jef-n @m-kuhn any objection?

File: tests/code_layout/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+ADD_TEST(qgis_sipify ${CMAKE_SOURCE_DIR}/tests/code_layout/test_sipify.sh)
+ADD_TEST(qgis_sip_uptodate ${CMAKE_SOURCE_DIR}/tests/code_layout/test_sipfiles_uptodate.sh)
+
+IF (WITH_APIDOC)
+  INCLUDE(UsePythonTest)
+  ADD_PYTHON_TEST(PyQgsDocCoverage ${CMAKE_SOURCE_DIR}/tests/code_layout/test_qgsdoccoverage.py)
+  IF (WITH_CORE)
+    ADD_PYTHON_TEST(PyQgsSipCoverage ${CMAKE_SOURCE_DIR}/tests/code_layout/test_qgssipcoverage.py)
+   ENDIF(WITH_CORE)
```
Comment (m-kuhn) [inline tests/code_layout/CMakeLists.txt]: Bad indentation

File: CMakeLists.txt [hunk 4]
```diff
@@ -34,298 +34,305 @@ SET(CMAKE_MODULE_PATH ${CMAKE_SOURCE_DIR}/cmake ${CMAKE_MODULE_PATH})
...
-
-  SET (WITH_SERVER_PLUGINS TRUE CACHE BOOL "Determines whether QGIS server support for python plugins should be built")
-  IF(WITH_SERVER_PLUGINS)
-    SET(HAVE_SERVER_PYTHON_PLUGINS TRUE)
-  ENDIF(WITH_SERVER_PLUGINS)
-ENDIF(WITH_SERVER)
-
-# Custom widgets
-SET (WITH_CUSTOM_WIDGETS FALSE CACHE BOOL "Determines whether QGIS custom widgets for Qt Designer should be built")
+SET (WITH_CORE TRUE CACHE BOOL "Determines whether QGIS core should be built.")
```
Comment (m-kuhn) [inline CMakeLists.txt]: Can you mark that as advanced? (`MARK_AS_ADVANCED(WITH_CORE)`)

**Comment by m-kuhn** [comment]
Looks good.
I like the move to place astyle to a separate place. I wonder if a directory 3rdparty or external would be more verbose than lib?
